data/transactions/logic/eval.go

```diff
@@ -4801,19 +4807,49 @@ func opItxnSubmit(cx *EvalContext) error {
 	}
 
 	ep := NewInnerEvalParams(cx.subtxns, cx)
+
+	if ep.Debugger != nil {
+		err := ep.Debugger.BeforeInnerTxnGroup(ep)
```

Comment:
I wasn't sure where to comment, as I think my comment is about a lot of the code.  This "interface for each method, lots of types switches, call if implemented" approach is pretty strange for Go.  What is leading you down this path? Why not a simple (though large) interface for a Debugger that has all these methods?

If it's about convenience for potential implementors, I would propose that you make a `NullDebugger` that implements the large interface, but with essentially empty implementations. Then, if someone wants to implements a Debugger that only does `BoforeInnerTxnGroup` and doesn't want to write a lot of code, they:

```
type TinySimpleDebugger {
   NullDebugger
}

func (tsd TinySimpleDebugger) BeforeInnerTxnGroup() {
  // stuff
}
```
Now, TinySimpleDebugger{NullDebugger{}} implements the whole interface, but was convenient to write.

Reply:
Good points. Originally we wanted optional hooks because `EvalContext.refreshDebugState` was called for every hook and was potentially expensive. Since that is no longer the case, I agree with your suggestions.

Changes made in e5fd2a1fc3ad7988fe7fe8a2ff3dcfe93524c002

Reply:
FWIW this concept is also the "I" in "SOLID". I'm not a zealot, just an observation.

Another option would be to have one or more debugger objects with the `interface{}` type and runtime checks to see what they do:
```
for _, debugger := range debuggers {
	if v, ok := (debugger).(debuggerAfterLogicEvalHook); ok {
		v.BeforeLogicEval(ctx)
	}
}
```

This gives you a lot of flexibility in the future (maybe more than you need).

I used this pattern in conduit recently: https://github.com/algorand/indexer/pull/1285/files